Extend error handling and tests of signal blocker

Martchus opened the pull request:
This may help to debug problems like https://github.com/os-autoinst/os-autoinst/issues/2549.

Not sure whether it makes sense to merge (all of) this. It might be helpful to hotpatch an affected worker before we merge this.

Martchus committed:
Extend error handling and tests of signal blocking

This may help to debug problems like
https://github.com/os-autoinst/os-autoinst/issues/2549.

Martchus commented:
I removed the

```
my @threads = split('\n', `grep ^SigBlk /proc/$$/task/*/status`);
my $expected_mask = sprintf('0%x', (1 << SIGCHLD - 1) | (1 << SIGTERM - 1));
like shift @threads, qr/0000000000000000/, 'signal mask of first thread is cleared';
like $_, qr/$expected_mask/, 'signal mask of other threads blocks SIGCHLD/SIGTERM' for @threads;
```

part because it is maybe a bit too specific for a unit test.

Not sure how useful this whole PR still is. It would definitely help to uncover cases in production where the signal blocker isn't working at all. However, following https://github.com/os-autoinst/os-autoinst/issues/2549 it seems that only a small number of threads are affected and we need to find out where they come from.

baierjan commented on ppmclibs/tinycv_impl.cc:
```diff
@@ -364,11 +368,34 @@ void create_opencv_threads(int thread_count)
+        const auto is_sigterm_blocked = sigismember(&current_set, SIGTERM);
+        if (is_sigchld_blocked != 1 || is_sigterm_blocked != 1) {
+            if (is_sigchld_blocked == -1 || is_sigterm_blocked == -1) {
+                auto lock = std::unique_lock<std::mutex>(m);
```
I guess this it the lock declared on https://github.com/os-autoinst/os-autoinst/pull/2551/files#diff-56ffbf3709dd0405661f7f4e7e2fafcf718a5f766d54f92c31f59c40590048c8L366?

Can you please explain quickly to me this assignment? It is not locking/unlocking the mutex here, or is it?

Martchus replied:
> It is not locking/unlocking the mutex here, or is it?

It *is* locking/unlocking the mutex here using c'tor/d'tor magic. It does that so the log messages are not mixed up.

baierjan replied:
Ah, right. That makes sense. It wasn't clear from the documentation on the first read (as it states the lock is optional, but I see it now). Thanks!

Martchus committed:
Avoid crashes due to potentially blocking on thread too less

In case `cv::getNumThreads()` is higher than the number of CPUs we don't
expect that many threads to be spawned but just `cv::getNumberOfCPUs()`
count. Not sure why I put the `-1` into commit
8363ce364ec381e3e45a72d2028c8caf609a6811 but it seems wrong.

I tested this by inserting `cv::setNumThreads(1000);` at the beginning of
`opencv_default_thread_count()` so that the limit by the number of CPUs is
actually effective. Then the signal block test only deadlocked with
`cv::getNumberOfCPUs() + 1` but just `cv::getNumberOfCPUs()` was fine. So
I think that is what the expected thread count should be.

One can also play around with this by invoking `t/28-signalblocker.t`
with `OS_AUTOINST_TEST_THREAD_COUNT=…`. For me it only deadlocks when
setting the variable to one more CPU threads that my system has but the
exact number of threads works.

Related issue: https://github.com/os-autoinst/os-autoinst/issues/2549

Martchus commented:
By the way, I added the "not-ready" label because the discussion on https://github.com/os-autoinst/os-autoinst/issues/2549 is still ongoing and removing the `-1` might be questionable.
